tuf/tuf.go

```diff
@@ -786,27 +820,148 @@ func (tr *Repo) UpdateTimestamp(s *data.Signed) error {
 	return nil
 }
 
-// SignRoot signs the root
+type versionedRootRole struct {
+	data.BaseRole
+	version int
+}
+
+type versionedRootRoles []versionedRootRole
+
+func (v versionedRootRoles) Len() int           { return len(v) }
+func (v versionedRootRoles) Swap(i, j int)      { v[i], v[j] = v[j], v[i] }
+func (v versionedRootRoles) Less(i, j int) bool { return v[i].version < v[j].version }
+
+// SignRoot signs the root, using all keys from the "root" role (i.e. currently trusted)
+// as well as available keys used to sign the previous version, if the public part is
+// carried in tr.Root.Keys and the private key is available (i.e. probably previously
+// trusted keys, to allow rollover).  If there are any errors, attempt to put root
+// back to the way it was (so version won't be incremented, for instance).
 func (tr *Repo) SignRoot(expires time.Time) (*data.Signed, error) {
 	logrus.Debug("signing root...")
-	tr.Root.Signed.Expires = expires
-	tr.Root.Signed.Version++
-	root, err := tr.GetBaseRole(data.CanonicalRootRole)
+	currRoot, err := tr.GetBaseRole(data.CanonicalRootRole)
 	if err != nil {
 		return nil, err
 	}
+
+	oldRootRoles, origRoles := tr.getOldRootRoles()
+
+	var latestSavedRole data.BaseRole
+	rolesToSignWith := make([]data.BaseRole, 0, len(oldRootRoles))
+
+	if len(oldRootRoles) > 0 {
+		sort.Sort(oldRootRoles)
+		for _, vRole := range oldRootRoles {
+			rolesToSignWith = append(rolesToSignWith, vRole.BaseRole)
+		}
+		latest := rolesToSignWith[len(rolesToSignWith)-1]
+		latestSavedRole = data.BaseRole{
+			Name:      data.CanonicalRootRole,
+			Threshold: latest.Threshold,
+			Keys:      latest.Keys,
+		}
+	}
+
+	// if the original role had not been saved as a previous role, save it now
+	if !tr.originalRootRole.Equals(latestSavedRole) {
+		tr.saveOldRootRole(tr.originalRootRole, tr.Root.Signed.Version)
+		rolesToSignWith = append(rolesToSignWith, tr.originalRootRole)
+		latestSavedRole = tr.originalRootRole
+	}
+
+	origVersion := tr.Root.Signed.Expires
+	tr.Root.Signed.Expires = expires
+	tr.Root.Signed.Version++
+
+	// if the current role doesn't match with the latest saved role, save it
+	if !currRoot.Equals(latestSavedRole) {
+		tr.saveOldRootRole(currRoot, tr.Root.Signed.Version)
+		rolesToSignWith = append(rolesToSignWith, currRoot)
+	}
+
 	signed, err := tr.Root.ToSigned()
 	if err != nil {
```

Comment:
rather than resetting, how complicated would it be to create a copy of the root and assign the whole thing over the top of `tr.Root` on success?